implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs

```diff
@@ -0,0 +1,314 @@
+            SendMessage(message) => {
+                info!(
+                    "remote route {:?}, decryptor remote {:?}",
+                    self.remote_route.clone(),
+                    self.addresses.decryptor_remote.clone()
+                );
+                context
+                    .send_from_address(
+                        self.remote_route(),
+                        message,
+                        self.addresses.decryptor_remote.clone(),
+                    )
+                    .await
+            }
+            Action::NoAction => Ok(()),
+        }
+    }
+
+    /// Handle a message coming from the other party
+    /// If the handshake has been fully performed then we can delegate this message to the
+    /// secure channel Decryptor.
+    /// Otherwise we unpack the message payload and send it to the state machine to trigger
+    /// a transition
+    async fn handle_message(
+        &mut self,
+        context: &mut Self::Context,
+        message: Routed<Self::Message>,
+    ) -> Result<()> {
+        if let Some(decryptor) = self.decryptor.as_mut() {
+            // since we cannot replace this worker with the decryptor worker we act as a relay instead
+            return decryptor.handle_message(context, message).await;
+        };
+
+        // we only set it once to avoid redirects attack
+        self.remote_route = Some(message.return_route());
+        if let SendMessage(message) = self
+            .state_machine
+            .on_event(ReceivedMessage(Vec::<u8>::decode(
+                &message.into_transport_message().payload,
+            )?))
+            .await?
+        {
+            context
+                .send_from_address(
+                    self.remote_route(),
+                    message,
+                    self.addresses.decryptor_remote.clone(),
+                )
+                .await?
+        };
+
+        // if we reached the final state we can make a pair of encryptor/decryptor
+        if let Some(final_state) = self.state_machine.get_handshake_results() {
+            // start the encryptor worker and return the decryptor
+            self.decryptor = Some(self.finalize(context, final_state).await?);
+            self.callback_sender.send(()).await?;
+        };
+        Ok(())
+    }
+}
+
+impl HandshakeWorker {
+    /// Create a new HandshakeWorker with a role of either INITIATOR or RESPONDER
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) async fn create(
+        context: &Context,
+        secure_channels: Arc<SecureChannels>,
+        addresses: Addresses,
+        identifier: IdentityIdentifier,
+        trust_policy: Arc<dyn TrustPolicy>,
+        decryptor_outgoing_access_control: Arc<dyn OutgoingAccessControl>,
+        credentials: Vec<Credential>,
+        trust_context: Option<TrustContext>,
+        remote_route: Option<Route>,
+        timeout: Option<Duration>,
+        role: Role,
+    ) -> Result<Address> {
```

Comment:
What is the point of returning an Address here?

Reply:
I removed it, it's indeed not needed.